parseAPI/h/CFG.h

```diff
@@ -510,8 +511,8 @@ class PARSER_EXPORT Function : public AnnotatableSparse, public boost::lockable_
     {
         bool operator()(const Function * f1, const Function * f2) const
         {
-            if (f1->region() < f2->region()) return true;
-            else if (f1->region() == f2->region() &&
+            if (f1->region()->low() < f2->region()->low()) return true;
```

Comment:
There are a few places in Dyninst that check if a CodeRegion exists before creating a Function object (e.g., [createAndRecordFunc](https://github.com/dyninst/dyninst/blob/master/parseAPI/src/ParseData.C#L139)), so it seems at least possible that a Function may not have an associated CodeRegion- whatever that means. In that case, I would just compare on the functions' addresses.

```suggestion
if(!(f1->region() && f2->region())) {
  return f1->addr() < f2->addr();
}

if (f1->region()->low() < f2->region()->low()) return true;
```

While we're here, let's also take the opportunity to clean up the code a bit.

```suggestion
auto *f1_region = f1->region();
auto *f2_region = f2->region();

if(!(f1_region && f2_region)) {
  return f1->addr() < f2->addr();
}

if(f1_region < f2_region) return true;

if(f1_region == f2_region && f1->addr() < f2->addr()) {
  return true;
}

return false;
```

Reply:
I've updated the code with your suggestion with a bit of alterations (no auto since it is not necessary here and it is easier to read when types are explicit; fewer `if`).

Reply:
>no auto since it is not necessary here and it is easier to read when types are explicit

Eh. I am an ardent (ab?)user of _almost always auto_, but opinions vary.

>fewer if

This allows one of the regions to be unspecified whereas the version I suggested requires both to be present. Let's err on the side of more verbose, but stricter.

Reply:
> This allows one of the regions to be unspecified whereas the version I suggested requires both to be present. Let's err on the side of more verbose, but stricter.

The code does exactly the same. `!(f1_region && f2_region) <=> (!f1_region || !f2_region)` Then I've just merged two `if` with the same body into one.

```
if (!f1_region || !f2_region) return f1->addr() < f2->addr();
if (f1_region->low() == f2_region->low()) return f1->addr() < f2->addr();
```

become

```
if (!f1_region || !f2_region || f1_region->low() == f2_region->low())
    return f1->addr() < f2->addr();
```

Reply:
>!(f1_region && f2_region) <=> (!f1_region || !f2_region)

Agreed. I wouldn't have used De Morgan here, but I'm not going to object.

>Then I've just merged two if with the same body into one.

Yes, but the second form allows short-circuiting evaluation of the equality check even if one of the regions is null.

Reply:
> Yes, but the second form allows short-circuiting evaluation of the equality check even if one of the regions is null.

Isn't it the same for the first form? For example, if f2_region is null:

f1_region && f2_region is false => !(f1_region && f2_region) is true => we return from the function and don't evaluate anything else.

Reply:
Oh, yes, sorry. I had the implicit pointer null check flipped in my head. Not enough coffee.